baybe/simulation/lookup.py

```diff
@@ -1,201 +1,301 @@
+            - ``"mean"``: Imputes the mean value for each target.
+            - ``"random"``: A random row will be used for the lookup.
 
     Raises:
-        AssertionError: If an analytical function is used and an incorrect number of
-            targets was specified.
+        ValueError: If an invalid lookup mechanism is provided.
+
+    Example:
+        >>> import pandas as pd
+        >>> from baybe.targets.numerical import NumericalTarget
+        >>> from baybe.simulation.lookup import look_up_targets
+        >>>
+        >>> targets = [NumericalTarget("target", "MAX")]
+        >>> df = pd.DataFrame({"x": [1, 2, 3]})
+        >>> lookup_df = pd.DataFrame({"x": [1, 2], "target": [10, 20]})
+        >>> look_up_targets(df, targets, lookup_df, impute_mode="mean")
+        >>> print(df)
+           x  target
+        0  1    10.0
+        1  2    20.0
+        2  3    15.0
     """
-    # TODO: This function needs another code cleanup and refactoring. In particular,
-    #   the different lookup modes should be implemented via multiple dispatch.
+    if lookup is None:
+        add_fake_results(queries, targets)
+    elif isinstance(lookup, Callable):
+        _look_up_targets_from_callable(queries, targets, lookup)
+    elif isinstance(lookup, pd.DataFrame):
+        _look_up_targets_from_dataframe(queries, targets, lookup, impute_mode)
+    else:
+        raise ValueError("Invalid lookup mechanism.")
 
-    # Extract all target names
-    target_names = [t.name for t in campaign.targets]
 
-    # If no lookup is provided, invent some fake results
-    if lookup is None:
-        add_fake_results(queries, campaign)
+def _look_up_targets_from_callable(
+    queries: pd.DataFrame,
+    targets: Collection[Target],
+    lookup: Callable,
+) -> None:
+    """Look up target values by querying a callable."""
+    # TODO: Currently, the alignment of return values to targets is based on the
+    #   column ordering, which is not robust. Instead, the callable should return
+    #   a dataframe with properly labeled columns.
 
-    # Compute the target values via a callable
-    elif isinstance(lookup, Callable):
-        # TODO: Currently, the alignment of return values to targets is based on the
-        #   column ordering, which is not robust. Instead, the callable should return
-        #   a dataframe with properly labeled columns.
-
-        # Since the return of a lookup function is a tuple, the following code stores
-        # tuples of floats in a single column with label 0:
-        measured_targets = queries.apply(lambda x: lookup(*x.values), axis=1).to_frame()
-        # We transform this column to a DataFrame in which there is an individual
-        # column for each of the targets....
-        split_target_columns = pd.DataFrame(
-            measured_targets[0].to_list(), index=measured_targets.index
+    # Since the return of a lookup function is a tuple, the following code stores
```

Comment:
If we are making explicit assumptions on what the lookup needs to look like, shouldn't we then have a more informative type hint then just `Callable`?

Reply:
might come in the next PR once we make the callable accept `pd.DataFrame`

Reply:
will be completely refactored in its own PR